labs/Demos/Caesar.md

```diff
@@ -0,0 +1,284 @@
+# Caesar Cipher
+
+The [Caesar cipher](https://en.wikipedia.org/wiki/Caesar_cipher) is a 
+simple shift cipher named after Julius Caesar. This cipher was...
+evidently effective against illiterate adversaries of ancient Rome, 
+but will it protect its secrets from you and Cryptol...?
+
+```
+/* Simple Caesar cipher */
+module labs::Demos::Caesar where
+```
+
+A Caesar cipher simply rotates characters in the alphabet by a fixed 
+amount. (Caesar himself used a left rotation by 3, according to 
+Suetonius.) We could define this simply and directly in Cryptol:
+
+```
+caesar msg = map rot3 msg
+  where
+    rot3 c =
+      if 'A' <= c /\ c <= 'Z' then
+        (['A'..'Z':Char] >>> 3) @ (c - 'A')
+      else
+        c
+```
+
+```sh
+labs::Demos::Caesar> :s ascii=on
+labs::Demos::Caesar> caesar "ATTACK AT DAWN"
+"XQQXZH XQ AXTK"
+```
+
+...and that's it! Thank you for trying the Caesar demo...
+
+...
+
+
+## Generalization
+
+OK, we can explore this a little further. Let's generalize it:
+
+```
+/** number of characters in alphabet */
+type w = 26
+
+/** an alphabet of characters included in a rotation cipher */
+type Alphabet = String w
+
+/** a left-rotation amount */
+type Key = [width w]
+
+/** alphabet of characters to "encrypt" */
+alphabet = ['A'..'Z'] : Alphabet
+```
+
+
+## Aside: A confounding search
+
+The above definition assumed a continguous character set, which does 
+not necessarily hold in general. So we'll need a function to find a 
+character in the alphabet:
+
+```
+/**
+ * index (from end) of first occurence (from start) of item `x` in 
+ * sequence `L`
+ */
+index L x = if (or M) then (lg2 ((0b0 # M) + 1) - 1) else (length M)
+  where
+    M = (map ((==) x) L)
+```
+
+It's...not obvious how this function works, so let's establish a 
+property to verify its correctness:
+
+```
+/** `index` is correct for any sequence */
+indexCorrect L x = (~ any ((==) x) L) \/ (L ! (index L x) == x)
```

Comment:
How about
```
indexCorrect L x = elem x L ==> L ! (index L x) == x
```
This reads a little clearer to me --- x being an element of L implies that ...

Reply:
Agreed. I forgot about `elem` being defined in the prelude. As for the `==>` form, it took me a moment to understand why it also covers the `~ (elem x L)` case lazily, but it does.